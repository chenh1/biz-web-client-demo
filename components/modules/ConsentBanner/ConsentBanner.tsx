"use client";

import React, { useState, useEffect, FC } from "react";
import { Section } from "../../containers/Section/Section";
import { Box } from "../../containers/Box/Box";
import { Button } from "../../core/Button/Button";
import { SpanLink } from "../../core/SpanLink/SpanLink";
import Link from "next/link";
import { Paragraph } from "../../core/Paragraph/Paragraph";
import { useCookies, Cookies } from "react-cookie";
import { useSession } from "next-auth/react";
import { usePathname } from "next/navigation";

export const CONSENT_COOKIE_NAME = "consent:accepted";

const rootPathsToRestrictAccess: string[] = ['/posts'];

export const ConsentBanner: FC = () => {
  const pathName = usePathname();
  const session = useSession();
  const [cookies, setCookie, removeCookie] = useCookies([CONSENT_COOKIE_NAME]);
  const [isReadyToShowBanner, setIsReadyToShowBanner] = useState<boolean>(false);
  const isRestrictedPage = rootPathsToRestrictAccess.some(path => pathName?.startsWith(path));

  useEffect(() => {
    if (typeof window !== 'undefined' && session?.status === 'unauthenticated') {
      setIsReadyToShowBanner(true);
    }
  }, [session?.status, pathName]);

  useEffect(() => {
    if (isRestrictedPage) {
      document.body.style.overflow = cookies[CONSENT_COOKIE_NAME] ? 'auto' : 'hidden';

      return () => {
        document.body.style.overflow = 'auto';
      };
    }
  }, [pathName, cookies[CONSENT_COOKIE_NAME], isRestrictedPage]);

  const hideBanner: boolean = !isReadyToShowBanner || !!cookies[CONSENT_COOKIE_NAME];

  return hideBanner ? null : (
    <>
      {isRestrictedPage && <div className="fixed top-0 w-screen h-screen z-50 bg-night-black opacity-25"></div>}
      <Section heightClass="py-12" className={`fixed bottom-0 ${isRestrictedPage ? 'h-1/2' : ''} w-screen bg-max-white z-50`}>
        <Box padding="0" gap="md" className="items-center sm:grid-cols-2">
          <Box padding="0" gap="sm">
            <Paragraph>We use cookies to enhance your user experience. By using our website, you agree to our use of cookies.</Paragraph>
            <Link href="/privacy-policy">
              <SpanLink uncolored>
                See our Privacy Policy
              </SpanLink>
            </Link>
          </Box>
          <Box padding="0" className="items-center">
            <Button onClick={() => setCookie(CONSENT_COOKIE_NAME, true)} primary label="Accept" size="lg" />
          </Box>
        </Box>
      </Section>
    </>
  );
};
