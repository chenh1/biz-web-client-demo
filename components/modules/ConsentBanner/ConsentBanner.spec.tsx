import { ConsentBanner } from "./ConsentBanner";
import { render, waitFor } from '@testing-library/react';
import { JSDOM } from "jsdom"
import { usePathname } from "next/navigation";
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

jest.mock('next/navigation', () => ({
  usePathname: () => "/",
}));

jest.mock('react-cookie', () => ({
  useCookies: jest.fn()
}));

jest.mock('next-auth/react', () => ({
  useSession: jest.fn()
}))

jest.mock('next/link', () => ({
  __esModule: true,
  default: ({ children }) => children
}));

import { useCookies } from "react-cookie";
import { useSession } from "next-auth/react";

const mockedUseCookies = useCookies as jest.Mock;
const mockedUseSession = useSession as jest.Mock;


it('displays the consent banner if unauthenticated and consent not given', async () => {
  mockedUseCookies.mockImplementation(() => [{ "consent:accepted": false }, jest.fn(), jest.fn()]);
  mockedUseSession.mockImplementation(() => ({ status: "unauthenticated" }));

  const { getByTestId, asFragment } = render(
    <ConsentBanner/>
  );

  await waitFor(() => new Promise((res) => setTimeout(res, 800)));

  expect(asFragment()).toMatchSnapshot();
});


it('hides the consent banner if consent accepted', async () => {
  mockedUseCookies.mockImplementation(() => [{ "consent:accepted": true }, jest.fn(), jest.fn()]);
  mockedUseSession.mockImplementation(() => ({ status: "unauthenticated" }));

  const { getByTestId, asFragment } = render(
    <ConsentBanner/>
  );

  await waitFor(() => new Promise((res) => setTimeout(res, 800)));

  expect(asFragment()).toMatchSnapshot();
});

it('hides the consent banner if user is authenticated', async () => {
  mockedUseCookies.mockImplementation(() => [{ "consent:accepted": false }, jest.fn(), jest.fn()]);
  mockedUseSession.mockImplementation(() => ({ status: "authenticated" }));

  const { getByTestId, asFragment } = render(
    <ConsentBanner/>
  );

  await waitFor(() => new Promise((res) => setTimeout(res, 800)));

  expect(asFragment()).toMatchSnapshot();
});
