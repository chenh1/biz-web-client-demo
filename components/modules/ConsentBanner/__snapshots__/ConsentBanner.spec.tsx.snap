// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays the consent banner if unauthenticated and consent not given 1`] = `
<DocumentFragment>
  <div
    class=" flex border-box justify-around w-full py-12 fixed bottom-0  w-screen bg-max-white z-50"
  >
    <div
      class="w-full max-w-7xl px-4 xl:px-0"
    >
      <div
        class=" grid  gap-8  p-2 items-center sm:grid-cols-2"
      >
        <div
          class=" grid  gap-4  p-2 "
        >
          <p
            class=" text-night-black dark:text-max-white font-body font-body text-base undefined"
          >
            We use cookies to enhance your user experience. By using our website, you agree to our use of cookies.
          </p>
          <span
            class="  relative cursor-pointer font-body text-night-black dark:text-max-white before:content-[""] before:absolute before:translate-y-[6px] before:bottom-0 before:duration-300 before:ease-in-out before:border-apex-blue before:w-0 before:border-2 before:opacity-0 hover:brightness-110 hover:before:w-full hover:before:opacity-100"
          >
            See our Privacy Policy
          </span>
        </div>
        <div
          class=" grid  gap-4  p-2 items-center"
        >
          <button
            class="   w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-max-white px-6 py-3   bg-hyper-red border-hyper-red dark:bg-hyper-red dark:border-hyper-red"
            data-testid="button"
            type="button"
          >
            <div
              class="bg-hyper-red border-hyper-red dark:bg-hyper-red dark:border-hyper-red rounded-lg absolute left-0 top-0 w-full h-full "
            />
            <span
              class="pointer-events-none relative z-10"
            >
              Accept
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`hides the consent banner if consent accepted 1`] = `<DocumentFragment />`;

exports[`hides the consent banner if user is authenticated 1`] = `<DocumentFragment />`;
